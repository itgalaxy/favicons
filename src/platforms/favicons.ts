import { FaviconHtmlElement } from "../index";
import { FaviconOptions, NamedIconOptions } from "../config/defaults";
import { transparentIcon, transparentIcons } from "../config/icons";
import { OptionalMixin, Platform, uniformIconOptions } from "./base";

const ICONS_OPTIONS: (NamedIconOptions & OptionalMixin)[] = [
  { name: "favicon.ico", ...transparentIcons(16, 24, 32, 48, 64) },
  { name: "favicon-16x16.png", ...transparentIcon(16) },
  { name: "favicon-32x32.png", ...transparentIcon(32) },
  { name: "favicon-48x48.png", ...transparentIcon(48) },
  { name: "favicon.svg", ...transparentIcon(1024), optional: true }, // arbitrary size. if more than one svg source is given, the closest to this size will be picked.
];

export class FaviconsPlatform extends Platform {
  constructor(options: FaviconOptions) {
    super(
      options,
      uniformIconOptions(options, options.icons.favicons, ICONS_OPTIONS),
    );
  }

  override async createHtml(): Promise<FaviconHtmlElement[]> {
    return this.iconOptions.map(({ name, ...options }) => {
      if (name.endsWith(".ico")) {
        // prettier-ignore
        return `<link rel="icon" type="image/x-icon" href="${this.cacheBusting(this.relative(name))}">`;
      } else if (name.endsWith(".svg")) {
        // prettier-ignore
        return `<link rel="icon" type="image/svg+xml" href="${this.cacheBusting(this.relative(name))}">`;
      }

      const { width, height } = options.sizes[0];

      // prettier-ignore
      return `<link rel="icon" type="image/png" sizes="${width}x${height}" href="${this.cacheBusting(this.relative(name))}">`;
    });
  }
}
