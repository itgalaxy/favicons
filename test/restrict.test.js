import favicons from "../src";
import { logo_png } from "./util";

const source = logo_png;

const options = {
  icons: {
    favicons: ["favicon-32x32.png"],
    android: false,
    appleIcon: false,
    appleStartup: false,
    windows: false,
    yandex: false,
  },
};

const metaLength = 0;

const expectedLength =
  Object.values(options.icons).reduce(
    (acc, elem) => acc + (Array.isArray(elem) ? elem.length : 0),
    0
  ) + metaLength;

function processSource() {
  return favicons(source, options);
}

function generatedHTML(html) {
  return html.filter((str) => str.length > 0);
}

test("should allow to restrict the icons to generate", async () => {
  // eslint-disable-next-line no-magic-numbers
  const testCases = [
    {
      // favicons
      source,
      options,
    },
    {
      // android
      source,
      options: {
        ...options,
        icons: {
          ...options.icons,
          favicons: false,
          android: ["android-chrome-48x48.png"],
        },
      },
    },
    {
      // appleIcon
      source,
      options: {
        ...options,
        icons: {
          ...options.icons,
          favicons: false,
          appleIcon: ["apple-touch-icon-60x60.png"],
        },
      },
    },
  ];

  expect.assertions(testCases.length);

  for (const { source, options } of testCases) {
    const { images } = await favicons(source, options);

    expect(images.length).toBe(expectedLength);
  }
});

test("should allow to restrict the HTML tags generated", async () => {
  // eslint-disable-next-line no-magic-numbers
  expect.assertions(1);

  const { html } = await processSource();

  const useful = generatedHTML(html);

  expect(useful.length).toBe(expectedLength);
});

test("should allow to restrict the HTML tags taking into account manifests and others", async () => {
  const testCases = [
    {
      // android
      testOptions: {
        favicons: false,
        android: ["android-chrome-48x48.png"],
        appleIcon: false,
        appleStartup: false,
        windows: false,
        yandex: false,
      },
      expectedLength: 4, // (manifest + capable + theme color + app name)
    },
    {
      // appleIcon
      testOptions: {
        favicons: false,
        android: false,
        appleIcon: ["apple-touch-icon-60x60.png"],
        appleStartup: false,
        windows: false,
        yandex: false,
      },
      expectedLength: 4, // icons + (capable + status bar + app name)
    },
  ];

  expect.assertions(testCases.length);

  for (const { testOptions, expectedLength } of testCases) {
    const { html } = await favicons(source, {
      icons: testOptions,
    });
    const useful = generatedHTML(html);

    expect(useful.length).toBe(expectedLength);
  }
});
