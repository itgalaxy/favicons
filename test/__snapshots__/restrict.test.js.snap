// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow to restrict the icons to generate and configure them simultaneously 3`] = `
{
  "files": [],
  "html": [
    "<link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">",
    "<link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">",
    "<meta name="apple-mobile-web-app-capable" content="yes">",
    "<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">",
    "<meta name="apple-mobile-web-app-title">",
  ],
  "images": [
    {
      "contents": null,
      "name": "apple-touch-icon-60x60.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-76x76.png",
    },
  ],
}
`;
